pkg/metrics/metrics.go

```diff
@@ -0,0 +1,118 @@
+/*
+Copyright 2018 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package metrics
+
+import (
+	"time"
+
+	"github.com/prometheus/client_golang/prometheus"
+	"k8s.io/klog"
+	"sigs.k8s.io/controller-runtime/pkg/metrics"
+)
+
+var (
+	kubefedClustersNotReadyCount = prometheus.NewCounter(
+		prometheus.CounterOpts{
+			Name: "kubefedclusters_not_ready_total",
+			Help: "Number of kubefed clusters not ready.",
+		},
+	)
+
+	kubefedClustersReadyCount = prometheus.NewCounter(
+		prometheus.CounterOpts{
+			Name: "kubefedclusters_ready_total",
+			Help: "Number of ready kubefed clusters.",
+		},
+	)
+
+	kubefedClustersOfflineCount = prometheus.NewCounter(
+		prometheus.CounterOpts{
+			Name: "kubefedclusters_offline_total",
+			Help: "Number of offline kubefed clusters.",
+		},
+	)
+
+	functionDuration = prometheus.NewHistogramVec(
+		prometheus.HistogramOpts{
+			Name:    "function_duration_seconds",
```

Comment:
This needs a better name - function is too generic and can apply to many different things, not necessarily just control loops. Maybe we can take inspiration from controller-runtime for this?

Reply:
That is why I pass the `function` as a way to distinguish between operations. Do you have any suggestion of a better name ?

Reply:
Take a look at https://github.com/kubernetes-sigs/controller-runtime/blob/c0438568a706ec61de31b92f4d76e7fb7e1007b9/pkg/internal/controller/metrics/metrics.go for inspiration perhaps?

Reply:
I'll update my PR this week. Thanks for the pointers